feat(tekton): view pipelinerun visualization on click of pipelinerun

debsmita1 opened the pull request:
**Resolves:**
https://github.com/janus-idp/backstage-plugins/issues/408

**Screenshots:**
<img width="1717" alt="Screenshot 2023-07-05 at 2 37 17 PM" src="https://github.com/janus-idp/backstage-plugins/assets/22490998/947dae0d-c05f-4cee-909e-933bd280263e">
<img width="1717" alt="Screenshot 2023-07-05 at 2 37 27 PM" src="https://github.com/janus-idp/backstage-plugins/assets/22490998/e6a62af1-a2c1-4abf-81b0-7f22d6e692bc">


https://github.com/janus-idp/backstage-plugins/assets/22490998/54afb29c-c404-4729-ba7e-09c92b1ed3ef

invincibleJai commented on plugins/tekton/src/plugin.ts:
```diff
@@ -4,34 +4,37 @@ import '@patternfly/patternfly/patternfly-charts-theme-dark.css';
 export const LatestPipelineRun = tektonPlugin.provide(
-  createRoutableExtension({
+  createComponentExtension({
     name: 'LatestPipelineRun',
```
should we rename this as well to `PipelineRun`

debsmita1 replied:
The LatestPipelineRun & PipelineRun are rendered on 2 different routes

invincibleJai commented on plugins/tekton/src/utils/pipelineRun-utils.ts:
```diff
@@ -97,6 +97,16 @@ export const getLatestPipelineRun = (
   return null;
 };
 
+export const getPipelineRun = (
```
can we have unit tests for this?

invincibleJai commented on plugins/tekton/src/components/pipeline-topology/PipelineVisualization.tsx:
```diff
@@ -147,12 +192,31 @@ export const PipelineVisualization = ({
       ) : (
         <div
-          data-testid="pipeline-visualization"
+          data-testid={
```
can we avoid this? the component should be declarative and ideally not good to change test id conditionally

invincibleJai commented on plugins/tekton/src/components/pipeline-topology/PipelineVisualization.tsx:
```diff
@@ -147,12 +192,31 @@ export const PipelineVisualization = ({
     </WrapperInfoCard>
   );
+
+  return pipelineRun ? (
```
this component has grown, can we refactor it?

invincibleJai commented on plugins/tekton/src/components/pipeline-topology/PipelineVisualization.test.tsx:
```diff
@@ -42,7 +48,40 @@ describe('PipelineVisualization', () => {
+
+    expect(screen.getByText('Back to PipelineRun list')).toBeInTheDocument();
+
+    expect(queryByTestId('pipelineRun-visualization')).not.toBeNull();
```
can be consistent with the above assertions in same it block

```suggestion
    expect(queryByTestId('pipelineRun-visualization')).toBeInTheDocument();;
```

invincibleJai commented on plugins/tekton/src/components/pipeline-topology/PipelineVisualization.test.tsx:
```diff
@@ -42,7 +48,40 @@ describe('PipelineVisualization', () => {
+    expect(queryByTestId('latest-pipelineRun-visualization')).not.toBeNull();
+  });
+
+  it('should render the pipeline run visualization when pipelineRun name exists in the params', async () => {
```
this doesn't need to be async

```suggestion
  it('should render the pipeline run visualization when pipelineRun name exists in the params', () => {
```

invincibleJai commented on plugins/tekton/src/components/pipeline-topology/PipelineVisualization.test.tsx:
```diff
@@ -16,8 +17,13 @@ jest.mock('@backstage/plugin-catalog-react', () => ({
+
 describe('PipelineVisualization', () => {
-  it('should render the pipeline run visualization when pipelineRun exists', async () => {
+  it('should render the latest pipeline run visualization when pipelineRun exists', async () => {
```
component should be declarative, and should just render PLR without the notion of this being the latest or old one.

invincibleJai commented on plugins/tekton/src/components/PipelineRunList/PipelineRunHeader.tsx:
```diff
@@ -20,11 +20,13 @@ export const PipelineRunHeader: TableColumn[] = [
-        abbr="PLR"
-        name={(row as PipelineRunKind)?.metadata?.name || ''}
-      />
+      <Link to={`pipelineRun/${(row as PipelineRunKind)?.metadata?.name}`}>
```
`(row as PipelineRunKind)?.metadata?.name` id duplicated here, good to keep it in const

invincibleJai commented on plugins/tekton/src/components/PipelineRunList/PipelineRunList.tsx:
```diff
@@ -64,7 +67,16 @@ const PipelineRunList = () => {
         showClusterSelector={clusters.length > 0}
       >
-        <EmptyState missing="data" title="No Pipeline Runs found" />
+        <Table
+          title="Pipeline Runs"
+          data={[]}
+          columns={PipelineRunHeader}
+          emptyContent={
+            <div data-testid="no-pipeline-runs" className={classes.empty}>
+              No Pipeline Runs found
+            </div>
+          }
+        />
```
do we need this or can we get rid of this if block itself?

invincibleJai commented on plugins/tekton/src/components/Tekton/index.ts:
```diff
@@ -1 +0,0 @@
-export { TektonComponent } from './TektonComponent';
```
we don't need this?

invincibleJai commented on plugins/tekton/src/components/pipeline-topology/PipelineVisualizationCard.test.tsx:
```diff
@@ -16,8 +17,41 @@ jest.mock('@backstage/plugin-catalog-react', () => ({
+      </TektonResourcesContext.Provider>,
+    );
+    expect(
+      screen.getByText(mockKubernetesPlrResponse.pipelineruns[1].metadata.name),
```
we can use `getByText` from the render 

`const { getByText } = render( ... )`

invincibleJai commented on plugins/tekton/src/components/pipeline-topology/PipelineVisualizationCard.test.tsx:
```diff
@@ -16,8 +17,41 @@ jest.mock('@backstage/plugin-catalog-react', () => ({
+      screen.getByText(mockKubernetesPlrResponse.pipelineruns[1].metadata.name),
+    ).toBeInTheDocument();
+
+    expect(queryByTestId('latest-pipelineRun-visualization')).not.toBeNull();
```
nit: 

```suggestion
    expect(queryByTestId('latest-pipelineRun-visualization')).toBeInTheDocument();
```

invincibleJai commented on plugins/tekton/src/plugin.ts:
```diff
@@ -4,34 +4,37 @@ import '@patternfly/patternfly/patternfly-charts-theme-dark.css';
 
 export const LatestPipelineRun = tektonPlugin.provide(
-  createRoutableExtension({
+  createComponentExtension({
```
curious what's the diff between `createRoutableExtension` and `createComponentExtension`?

invincibleJai commented:
in the case of multiple clusters with the first cluster having no PLR and the second having few and I select PLR from the list view on cluster 2 then I see empty PLR viz as context is missing.

debsmita1 commented:
> in the case of multiple clusters with the first cluster having no PLR and the second having few and I select PLR from the list view on cluster 2 then I see empty PLR viz as context is missing.

Fixed it ! Added Gif in the PR description

invincibleJai commented on plugins/tekton/src/routes.ts:
```diff
@@ -1,5 +1,11 @@
+export const pipelineRunRouteRef = createSubRouteRef({
+  id: 'pipeline-run-visualization',
+  parent: rootRouteRef,
+  path: '/cluster/:clusterName/pipelineRun/:pipelineRunName',
```
here it says `:clusterName` but what's passed in number i.e index https://github.com/janus-idp/backstage-plugins/pull/506/files#diff-cd65d6b98bf0c3daff8d2edaaad8e013f7a55c35d6df69374fca9a4c571c710eR28

debsmita1 replied:
`clusters[selectedCluster as number]` is the name of the selected cluster

invincibleJai replied:
ah my bad, thanks 👍

invincibleJai commented:
in dev mode am not able to click on PLR to see the PLR Viz 

<img width="1789" alt="image" src="https://github.com/janus-idp/backstage-plugins/assets/5129024/3fdc92a7-ad55-4120-a82f-9a63c79c9768">

invincibleJai reviewed (approved):
Thanks @debsmita1, this looks good.
